src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -371,18 +374,38 @@ void GcsAutoscalerStateManager::HandleDrainNode(
     return;
   }
 
+  auto node = std::move(maybe_node.value());
+
+  // Set the death reason of the node.
+  auto death_info = node->mutable_death_info();
```

Comment:
We should not directly access mutable field of node here. It is dangerous. 

Instead consider creating an node manager API to update the death info